File: Universal-GenAI-Agent/src/web-scraper.py
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.chrome.options import Options
from webdriver_manager.chrome import ChromeDriverManager
import requests

# Set up the Chrome options
chrome_options = Options()
chrome_options.add_argument("--headless")
chrome_options.add_argument("--disable-gpu")
chrome_options.add_argument("--no-sandbox")
chrome_options.add_argument("blink-settings=imagesEnabled=false")
chrome_options.add_argument("--disable-extensions")
chrome_options.add_argument("--disable-infobars")
chrome_options.add_argument("--disable-dev-shm-usage")
chrome_options.page_load_strategy = 'eager'

# Initialize the WebDriver
driver = webdriver.Chrome(service=Service(ChromeDriverManager().install()), options=chrome_options)

try:
    driver.get('https://www.scmp.com/')
    driver.implicitly_wait(5)
    scraped_content = driver.find_element(By.TAG_NAME, 'body').text

finally:
    driver.quit()

gemini_connector_url = 'http://localhost:3000/stream'
response = requests.post(gemini_connector_url, json={'webContent': scraped_content}, stream=True)

for line in response.iter_lines():
    if line:
        print(line.decode('utf-8'))

File: Universal-GenAI-Agent/src/web-scraper.js
// src\web-scraper.js
const { Builder, By, until } = require('selenium-webdriver');
const chrome = require('selenium-webdriver/chrome');

async function webScraper(url) {

    const options = new chrome.Options();
    options.addArguments('--headless');
    options.addArguments('--disable-gpu');
    options.addArguments('--no-sandbox');
    options.addArguments('--disable-dev-shm-usage');
    options.addArguments('--ignore-certificate-errors');

    // Initialize the Chrome driver
    let driver = await new Builder()
        .forBrowser('chrome')
        .setChromeOptions(options)
        .build();

    try {
        // Navigate to the specified URL
        await driver.get(url);

        // Wait for content to load
        await driver.wait(until.elementLocated(By.css('body')), 10000);

        console.log('Extracting content...');
        const content = await driver.executeScript(`
            function extractText(element) {
                if (element.tagName === 'SCRIPT' || element.tagName === 'STYLE') {
                    return '';
                }
                
                let text = '';
                for (let child of element.childNodes) {
                    if (child.nodeType === 3) { // Text node
                        text += child.textContent.trim() + ' ';
                    } else if (child.nodeType === 1) { // Element node
                        if (window.getComputedStyle(child).display !== 'none') {
                            text += extractText(child) + ' ';
                        }
                    }
                }
                return text.trim();
            }
            return extractText(document.body);
        `);

        if (!content) {
            throw new Error('No content extracted');
        }

        console.log('Content length:', content.length);
        return content;

    } catch (error) {
        console.error('Error during scraping:', error);
        throw error; // Rethrow the error to handle it in the calling function
    } finally {
        // Quit the driver
        await driver.quit();
    }
}

// Export the function
module.exports = webScraper;